Add functions for interacting with IMAS data

eldond opened the pull request:
New features:
- `geqdsk2imas!()` writes data from one or more GEQDSKFile instance to the IMAS schema
- `imas2geqdsk()` reads equilibrium and wall data from an IMAS dd and outputs GEQDSKFile(s)

Updates:
- The path is split from filenames when recording file into GEQDSKFile during io

Close #15

eldond committed:
Add functions for transferring geqdsk data to IMAS

eldond committed:
Add function for writing equilibrium top level data as well as time slice

eldond committed:
Add functions for adding a vector of geqdsk instances to IMAS

eldond committed:
Don't record the full file path into GEQDSK during readg

eldond committed:
Add method for going from imas to geqdsk

orso82 commented:
@eldond this looks good, though I have not tested it.

One thing you could try is to use the `identify_cocos()` method https://github.com/ProjectTorreyPines/CoordinateConventions.jl/blob/290e6b7e062cebcb4b07d422d39718271ad042d0/src/CoordinateConventions.jl#L304 since gEQDSKs can be in COCOS 1,3,5,7

There's an open issue about https://github.com/ProjectTorreyPines/CoordinateConventions.jl/issues/1 but that should not affect you here

The other thing is that I have now made `https://github.com/ProjectTorreyPines/IMAS.jl` public. I much rather you use the `IMAS.flux_surfaces(eqt)` function to get the derived quantities in `dd`. The intention is to register `IMAS.jl` in the general registry, though we need to wait a couple of days for the packages that it depends on to be registered there.

eldond commented:
@orso82 I deleted derived quantities other than X-points, since EFIT.jl has an X-point finding function. If IMAS can do better at finding X-points, then maybe EFIT's X-point finder should be deleted and IMAS should be used for this instead. If EFIT should keep an X-point finding function, then I think it's okay for its results to be loaded into IMASdd.

I will look at identify_cocos

eldond committed:
Add COCOS identification in case geqdsk_cocos is given as 0

eldond committed:
Add/update docstrings

eldond committed:
Split out generic part of geqdsk2imas doc and include in each function variant

eldond commented:
@orso82 Okay, I think this is done. I'm using identify_cocos, I put in a bunch of tests, and I documented the functions.

orso82 commented on src/io.jl:
```diff
@@ -223,9 +226,530 @@ function readg(gfile; set_time=nothing)
+(both of these needing instruction on which time slice to write to) or to a single
+equilibrium.time\\_slice. If the top level DD isn't referenced, a reference to the
+wall IDS can be passed in separately to allow writing wall data.
+The COCOS of the incoming geqdsk files can be spefificed, or it can be automatically
+detected by setting geqdsk\\_cocos=0 (the default). If multiple COCOS are possible,
+the first one in the list will be chosen, which usually means 3 intead of 13, for example.
+To narrow down the choice of COCOS, it is necessary to specify the direction of toroidal
+angle using cocos\\_clockwise\\_phi, which defaults to false (meaning phi is CCW).
```
@eldond gEQDSKs can be in either COCOS 1,3,5,7 depending on the direction of the current and magnetic field. Please take a look at https://github.com/gafusion/OMFIT-source/blob/92a14adeca963b6fe58520cd61f536be1ddee587/omfit/omfit_classes/omfit_eqdsk.py#L3263-L3293

I feel the `geqdsk2imas_generic` text does not add much to what's already in the docstrings without that text... no? Also, be conscious that the docstrings should look nice in MarkDown format, so maybe something like this format would look nicer when rendered in the online documentation:
```
    geqdsk2imas!(
        gs::Vector{GEQDSKFile},
        dd::IMASdd.dd;
        geqdsk_cocos::Int=0,
        dd_cocos::Int=11,
        cocos_clockwise_phi::Bool=false,
        add_derived::Bool=false)

Utility for writing equilibrium data from GEQDSKFiles to IMAS schema.

* gs: Vector of GEQDSKFile instances
* dd: Top level IMAS data dictionary
* geqdsk_cocos: coordinate convention indentifier of the incoming GEQDSK set. 0 to auto detect.
* dd_cocos: coordinate convention identifier of the DD. Should normally be left at 11.
* cocos_clockwise_phi: hint for deciding COCOS, because GEQDSK probably doesn't constrain this.
* add_derived: switch to use functions within EFIT.jl to extend equilibrium data
```

Notice that the docstring should not have `function geqdsk2imas!`and just `geqdsk2imas!` suffices

eldond replied:
Okay, take out unneeded function prefix (oops), change to bullets, and delete generic thing.

orso82 commented on src/io.jl:
```diff
@@ -223,9 +226,530 @@ function readg(gfile; set_time=nothing)
+    end
+    tc = CoordinateConventions.transform_cocos(geqdsk_cocos, dd_cocos)
+
+    # Make sure top level stuff is defined
```
I would suggest leaning on `IMASdd`'s time handling capabilities.  I do need to document these better, so that people are more aware!

A `dd` defines a `global_time` at which all time-dependent operations take place. In your case you should do:
```julia
IMAS.top_dd(eq).global_time = g.time # set the global_time based on the gEQKDS time

resize(eq.time_slice) # this resizes time_slice (if neeed) and sets eq.time_slice[].time (note that in IMAS eq.time is an IMAS expression, but IMASdd does not have expressions)

eqt = eq.time_slice[] # this gives you the time_slice at the current global_time

@ddtime(eq.vacuum_toroidal_field.b0 = g.bcentr .* tc["B"]) # this macro detects that `b0` is a time array and inserts data at the current global_time
```

If you want to be nice you should restore the original `global_time` that was set prior entering your functions.
```julia
dd = IMAS.top_dd(eq)
original_global_time = dd.global_time
try
   dd.global_time = g.time
   ...
finally
   dd.global_time = original_global_time
end
```

eldond replied:
@orso82 Thanks

orso82 commented on src/io.jl:
```diff
@@ -223,9 +226,530 @@ function readg(gfile; set_time=nothing)
+    return [imas2geqdsk(dd, time_index, geqdsk_cocos=geqdsk_cocos, dd_cocos=dd_cocos) for time_index in 1:nt]
+end
+
+"""
```
```suggestion
"""
    imas2geqdsk(
        dd::IMASdd.dd,
        time_index::Int;
        geqdsk_cocos::Int=1,
        dd_cocos::Int=11,
    )::GEQDSKFile
```

orso82 commented on test/runtests.jl:
```diff
@@ -94,3 +96,133 @@ end
 end
+
+function gcompare(g1::GEQDSKFile, g2::GEQDSKFile; note=nothing, verbose::Bool=false)
+    fields = split("file,time,nw,nh,r,z,rdim,zdim,rleft,zmid,nbbbs,rbbbs,zbbbs,limitr,rlim,zlim,"*
+    "rcentr,bcentr,rmaxis,zmaxis,simag,sibry,psi,current,fpol,pres,ffprim,pprime,"*
+    "qpsi,psirz,rhovn", ",")
```
```suggestion
    fields = fieldnames(GEQDSKFile)
```

orso82 commented on src/io.jl:
```diff
@@ -223,9 +226,530 @@ function readg(gfile; set_time=nothing)
+        geqdsk_cocos = CoordinateConventions.identify_cocos(
+            sign(g.bcentr), sign(g.current), sign(g.qpsi[1]), sign(g.psi[end] - g.psi[1]), cocos_clockwise_phi,
+        )[1]
+        println("  identified COCOS=$geqdsk_cocos")
```
Please don't pollute peoples' stdout
```suggestion
        @debug "  identified COCOS=$geqdsk_cocos"
```

orso82 reviewed (commented):
mostly comments on stile and using IMAS(dd) time handling features

eldond committed:
Adjust docstrings for imas-related functions

- input list as bullets
- don't put "function" before everything in doc
- cut excessive generic text for geqdsk2imas

eldond committed:
Don't manually list GEQDSKFile fieldnames for testing

Co-authored-by: Orso Meneghini <orso82@gmail.com>

eldond committed:
Change print statements to debug output in src/io.jl

Co-authored-by: Orso Meneghini <orso82@gmail.com>

eldond committed:
Try to use IMAS time handling instead of passing around indices

eldond committed:
Adapt tests to new time handling and fix problems

anchal-physics commented:
This is not related to this branch's main motive but it would be nice if this small fix can be piggy backed into it. At this line:
https://github.com/JuliaFusion/EFIT.jl/blob/60b41dce2d606ee5f21f00a5b843d78951f6e473/src/utils.jl#L194

Using `LinearInterpolations`is deprecated and will be removed in future. This warning is printed every ime we use this function. Please just rename it to `linear_interpolations`. Thanks.

eldond committed:
Update LinearInteroplation to linear_interpolation

orso82 commented on src/io.jl:
```diff
@@ -223,9 +226,545 @@ function readg(gfile; set_time=nothing)
+        g, dd.equilibrium;
+        wall=dd.wall, geqdsk_cocos=geqdsk_cocos, dd_cocos=dd_cocos, cocos_clockwise_phi=cocos_clockwise_phi,
+        add_derived=add_derived,
+    )
```
Not critical, but did you know that when the variable name matches the keyword argument name you can skip the keyword argument? Like this:
```suggestion
    geqdsk2imas!(
        g, dd.equilibrium;
        dd.wall, geqdsk_cocos, dd_cocos, cocos_clockwise_phi, add_derived,
    )
```

orso82 commented on src/io.jl:
```diff
@@ -223,9 +226,545 @@ function readg(gfile; set_time=nothing)
+    finally
+        dd.global_time = original_global_time
+    end
+end
```
```suggestion
    dd = IMASdd.top_dd(eq)
    original_global_time = dd.global_time
    try
        dd.global_time = g.time
        resize!(eq.time_slice)
        geqdsk2imas!(
            g, eq.time_slice[];
            wall, geqdsk_cocos, dd_cocos, cocos_clockwise_phi, add_derived,
        )
    finally
        dd.global_time = original_global_time
    end
end
```

orso82 commented on src/io.jl:
```diff
@@ -223,9 +226,545 @@ function readg(gfile; set_time=nothing)
+
+    if ismissing(eqt, :time)
+        eqt.time = g.time
+    end
```
unecessary
```suggestion
```

orso82 commented on src/io.jl:
```diff
@@ -223,9 +226,545 @@ function readg(gfile; set_time=nothing)
+        dd.global_time = g.time
+        if ismissing(dd.equilibrium.vacuum_toroidal_field, :b0)
+            dd.equilibrium.vacuum_toroidal_field.b0 = zeros(length(dd.equilibrium.time))
+        end
```
unnecessary
```suggestion
```

orso82 reviewed (commented):
👍 some more comments, mostly removing unnecessary manual handling of time-dependent quantities. Also some tips about keyword arguments.

@bclyons12 could you please chime in about `cocos_clockwise_phi` ?

eldond committed:
Optimize time handling and other IMASdd interactions

Co-authored-by: Orso Meneghini <orso82@gmail.com>

eldond commented:
Removing some of the time assignments / setup seems to have been problematic. @orso82 , the way I tried to resolve this previously may not have been well aligned with ideal IMAS usage. How would you recommend addressing this?
```
julia> include("test/runtests.jl")
Precompiling EFIT
  1 dependency successfully precompiled in 5 seconds. 123 already precompiled.
Test Summary: | Pass  Total  Time
find xpoint   |   20     20  9.7s
Got the expected exception
Test Summary: | Pass  Total  Time
parse header  |    4      4  0.1s
Test files are identified with COCOS = [5, 7]
test file 1 info: filename = g000001.01000, time = 0.001
imas: Error During Test at /home/eldond/.julia/dev/EFIT/test/runtests.jl:156
  Test threw exception
  Expression: eqt.time == (gs[1]).time
  equilibrium.time_slice[1].time is missing
  Stacktrace:
   [1] getproperty(ids::IMASdd.IDS, field::Symbol)
     @ IMASdd ~/.julia/dev/IMASdd/src/data.jl:213
   [2] macro expansion
     @ ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:669 [inlined]
   [3] macro expansion
     @ ~/.julia/dev/EFIT/test/runtests.jl:156 [inlined]
   [4] macro expansion
     @ ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
   [5] top-level scope
     @ ~/.julia/dev/EFIT/test/runtests.jl:143
test file 2 info: filename = g184833.03600, time = 3.6
imas: Error During Test at /home/eldond/.julia/dev/EFIT/test/runtests.jl:142
  Got exception outside of a @test
  equilibrium.time_slice[1].time is missing
  Stacktrace:
    [1] getproperty
      @ ~/.julia/dev/IMASdd/src/data.jl:213 [inlined]
    [2] time_array_local(ids::IMASdd.IDSvector{T} where T<:IMASdd.IDSvectorTimeElement)
      @ IMASdd ~/.julia/dev/IMASdd/src/time.jl:120
    [3] resize!(ids::IMASdd.IDSvector{IMASdd.equilibrium__time_slice{Float64}}, time0::Float64)
      @ IMASdd ~/.julia/dev/IMASdd/src/data.jl:969
    [4] resize!(ids::IMASdd.IDSvector{IMASdd.equilibrium__time_slice{Float64}})
      @ IMASdd ~/.julia/dev/IMASdd/src/data.jl:965
    [5] geqdsk2imas!(g::GEQDSKFile, eq::IMASdd.equilibrium{Float64}; wall::Nothing, geqdsk_cocos::Int64, dd_cocos::Int64, cocos_clockwise_phi::Bool, add_derived::Bool)
      @ EFIT ~/.julia/dev/EFIT/src/io.jl:427
    [6] macro expansion
      @ ~/.julia/dev/EFIT/test/runtests.jl:161 [inlined]
    [7] macro expansion
      @ ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
    [8] top-level scope
      @ ~/.julia/dev/EFIT/test/runtests.jl:143
    [9] include(fname::String)
      @ Base.MainInclude ./client.jl:489
   [10] top-level scope
      @ REPL[2]:1
   [11] eval
      @ ./boot.jl:385 [inlined]
   [12] eval_user_input(ast::Any, backend::REPL.REPLBackend, mod::Module)
      @ REPL ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:150
   [13] repl_backend_loop(backend::REPL.REPLBackend, get_module::Function)
      @ REPL ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:246
   [14] start_repl_backend(backend::REPL.REPLBackend, consumer::Any; get_module::Function)
      @ REPL ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:231
   [15] run_repl(repl::REPL.AbstractREPL, consumer::Any; backend_on_current_task::Bool, backend::Any)
      @ REPL ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:389
   [16] run_repl(repl::REPL.AbstractREPL, consumer::Any)
      @ REPL ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/REPL/src/REPL.jl:375
   [17] (::Base.var"#1013#1015"{Bool, Bool, Bool})(REPL::Module)
      @ Base ./client.jl:432
   [18] #invokelatest#2
      @ ./essentials.jl:892 [inlined]
   [19] invokelatest
      @ ./essentials.jl:889 [inlined]
   [20] run_main_repl(interactive::Bool, quiet::Bool, banner::Bool, history_file::Bool, color_set::Bool)
      @ Base ./client.jl:416
   [21] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:333
   [22] _start()
      @ Base ./client.jl:552
Test Summary: | Pass  Error  Total  Time
imas          |    1      2      3  6.9s
ERROR: LoadError: Some tests did not pass: 1 passed, 0 failed, 2 errored, 0 broken.
in expression starting at /home/eldond/.julia/dev/EFIT/test/runtests.jl:142
```

eldond commented:
@orso82 whoops, this got forgotten. Currently stuck on trying to use IMAS time handling as intended.

orso82 commented:
let me know if you need help @eldond

orso82 commented:
@eldong FYI that @mgyoo86 is interested in this

orso82 committed:
Merge remote-tracking branch 'origin/imas'

orso82 committed:
move IMAS things to separate file

orso82 commented:
@eldond @lstagner @anchal-physics FYI that IMAS.jl is now available in the General registry so we should now be able to merge this PR, and release a new public version of EFIT.jl

lstagner commented:
Sure. @eldond just let me know if there is anything else that needs to be done and I'll merge it.

orso82 committed:
compat IMASdd = "2,3"

bclyons12 committed:
Merge branch 'master' into imas
